### PR TITLE
[Optimize-Instructions] simplify sign patterns like x < 0 ? -1 : 1

### PR DESCRIPTION
**Pull request by MaxGraey**
`i32(x) < 0 ? i32(-1) : i32(1)` -> `x >> 31 | 1`
`i64(x) < 0 ? i64(-1) : i64(1)` -> `x >> 63 | 1`

this shrink 2 bytes and it btw didn't apply by LLVM

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzer found issue with seed `17999754320937199714` but I can't reduce this to something useful due this error:
```
Missing type: (func (param externref f64 funcref i64 f64 f64 i32) (result f64))
Assertion failed: (0), function getTypeIndex, file .../binaryen/src/wasm/wasm-binary.cpp, line 529.

[1]    1769 abort      .../binaryen/bin/wasm-reduce   -t  -w 
```

**Comment by kripken** [comment]
> 17999754320937199714

I think that assertion was fixed by #3752 - which just landed, so maybe you don't have it merged here yet?

**Comment by MaxGraey** [comment]
can confirm. this fixed

**Comment by MaxGraey** [comment]
Could I disable GC stuffs in fuzzer for now?

**Comment by kripken** [comment]
There should be very little GC being fuzzed right now (there are various `--disable-gc` in the fuzzer script. I don't think you should see any bugs. If you do let me know.

**Comment by MaxGraey** [comment]
I got this one `seed: 14904800870670691178` but I'm not sure is it relate to GC or not.

**Comment by kripken** [comment]
I see, thanks. That's https://github.com/WebAssembly/binaryen/issues/3759 which is more common than I thought I guess. I landed https://github.com/WebAssembly/binaryen/pull/3760 now which works around it, with that you shouldn't see that issue any more.

**Comment by MaxGraey** [comment]
Thanks! Another strange issue: `seed 5105565690520908449`

```
!
CompareVMs between VMs: binaryen interpreter and wasm2c comparison error, expected to have '[fuzz-exec] note result: func_23 => -114
[fuzz-exec] calling func_23_invoker
[fuzz-exec] calling func_25
[LoggingExternalInterface logging 0 0]
[fuzz-exec] calling func_26
[LoggingExternalInterface logging 0]' == '[fuzz-exec] note result: func_23 => -1.14000000000000000e+02
[fuzz-exec] calling func_23_invoker
[fuzz-exec] calling func_25
[LoggingExternalInterface logging 0]
[fuzz-exec] calling func_26
[LoggingExternalInterface logging 0]', diff:

--- expected

+++ actual

@@ -1,6 +1,6 @@

-[fuzz-exec] note result: func_23 => -114
+[fuzz-exec] note result: func_23 => -1.14000000000000000e+02
 [fuzz-exec] calling func_23_invoker
 [fuzz-exec] calling func_25
-[LoggingExternalInterface logging 0 0]
+[LoggingExternalInterface logging 0]
 [fuzz-exec] calling func_26
 [LoggingExternalInterface logging 0]
```

And it want't reducing.

**Comment by kripken** [comment]
That `0 0` vs `0` issue should have been fixed by https://github.com/WebAssembly/binaryen/pull/3747 (which landed in the last day). I don't see it happen on `main` now.

(btw you mentioned `master`, and we switched from that to `main` - maybe you're on older code because of that?)

**Comment by MaxGraey** [comment]
Ah, it was `main`. just out of habit called as master

**Comment by MaxGraey** [comment]
Fuzzed until `ITERATION: 16542`. 
@kripken @tlively сould you take a look at this PR when you have time?

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -1481,6 +1482,29 @@ struct OptimizeInstructions
...
+      Binary* bin;
+      if (matches(
+            curr,
+            select(ival(-1), ival(1), binary(&bin, LtS, any(), ival(0)))) ||
+          matches(
+            curr,
+            select(ival(1), ival(-1), binary(&bin, GeS, any(), ival(0))))) {
+        auto c = bin->right->cast<Const>();
+        auto type = curr->ifTrue->type;
+        if (type == c->type) {
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:1498]: Please add a testcase for the case where this is not true.

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1498]: Done!

**Review by kripken — APPROVED** [review]
Thanks!

**Pull request merged**